### PR TITLE
#1278 - Flaky TestBasicFlowsWithOriginatingIdentity

### LINKED ISSUE #1278: TestBasicFlowsWithOriginatingIdentity go test -race detected
in https://service-catalog-jenkins.appspot.com/job/service-catalog-PR-testing2/2249/console

```
[Build & Unit/Integration Tests] === RUN   TestBasicFlowsWithOriginatingIdentity
[Build & Unit/Integration Tests] [restful] 2017/09/26 14:47:10 log.go:33: [restful/swagger] listing is available at https:///swaggerapi
[Build & Unit/Integration Tests] [restful] 2017/09/26 14:47:10 log.go:33: [restful/swagger] https:///swaggerui/ is mapped to folder /swagger-ui/
[Build & Unit/Integration Tests] E0926 14:47:12.429886    8780 controller_instance.go:1171] Failed to update status for ServiceInstance test-namespace/test-instance: Operation cannot be fulfilled on serviceinstances "test-instance": the object has been modified; please apply your changes to the latest version and try again
[Build & Unit/Integration Tests] ==================
[Build & Unit/Integration Tests] WARNING: DATA RACE
[Build & Unit/Integration Tests] Write at 0x00c420214b10 by goroutine 1034:
[Build & Unit/Integration Tests]   runtime.mapassign_faststr()
[Build & Unit/Integration Tests]       /usr/local/go/src/runtime/hashmap_fast.go:598 +0x0
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/apiserver/pkg/util/feature.(*featureGate).Set()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/apiserver/pkg/util/feature/feature_gate.go:136 +0x28c
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/test/integration.TestBasicFlowsWithOriginatingIdentity()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/test/integration/controller_test.go:946 +0x29f6
[Build & Unit/Integration Tests]   testing.tRunner()
[Build & Unit/Integration Tests]       /usr/local/go/src/testing/testing.go:746 +0x16c
[Build & Unit/Integration Tests] 
[Build & Unit/Integration Tests] Previous read at 0x00c420214b10 by goroutine 1145:
[Build & Unit/Integration Tests]   runtime.mapaccess2_faststr()
[Build & Unit/Integration Tests]       /usr/local/go/src/runtime/hashmap_fast.go:315 +0x0
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/apiserver/pkg/util/feature.(*featureGate).Enabled()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/apiserver/pkg/util/feature/feature_gate.go:182 +0x11b
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/pkg/controller.(*controller).reconcileServiceInstanceCredential()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/pkg/controller/controller_binding.go:275 +0x4ad9
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/pkg/controller.(*controller).reconcileServiceInstanceCredentialKey()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/pkg/controller/controller_binding.go:66 +0x2d8
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/pkg/controller.(*controller).(github.com/kubernetes-incubator/service-catalog/pkg/controller.reconcileServiceInstanceCredentialKey)-fm()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/pkg/controller/controller.go:178 +0x55
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/pkg/controller.worker.func1.1()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/pkg/controller/controller.go:209 +0xf7
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/pkg/controller.worker.func1()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/pkg/controller/controller.go:226 +0xb2
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:97 +0x6f
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:98 +0xcd
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/apimachinery/pkg/util/wait.Until()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:52 +0x5a
[Build & Unit/Integration Tests] 
[Build & Unit/Integration Tests] Goroutine 1034 (running) created at:
[Build & Unit/Integration Tests]   testing.(*T).Run()
[Build & Unit/Integration Tests]       /usr/local/go/src/testing/testing.go:789 +0x568
[Build & Unit/Integration Tests]   testing.runTests.func1()
[Build & Unit/Integration Tests]       /usr/local/go/src/testing/testing.go:1004 +0xa7
[Build & Unit/Integration Tests]   testing.tRunner()
[Build & Unit/Integration Tests]       /usr/local/go/src/testing/testing.go:746 +0x16c
[Build & Unit/Integration Tests]   testing.runTests()
[Build & Unit/Integration Tests]       /usr/local/go/src/testing/testing.go:1002 +0x521
[Build & Unit/Integration Tests]   testing.(*M).Run()
[Build & Unit/Integration Tests]       /usr/local/go/src/testing/testing.go:921 +0x206
[Build & Unit/Integration Tests]   main.main()
[Build & Unit/Integration Tests]       github.com/kubernetes-incubator/service-catalog/test/integration/_test/_testmain.go:70 +0x1d3
[Build & Unit/Integration Tests] 
[Build & Unit/Integration Tests] Goroutine 1145 (running) created at:
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/pkg/controller.(*controller).Run()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/pkg/controller/controller.go:178 +0x6d6
[Build & Unit/Integration Tests] ==================
[Build & Unit/Integration Tests] ==================
[Build & Unit/Integration Tests] WARNING: DATA RACE
[Build & Unit/Integration Tests] Write at 0x00c424c9ac68 by goroutine 1034:
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/apiserver/pkg/util/feature.(*featureGate).Set()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/apiserver/pkg/util/feature/feature_gate.go:136 +0x2a5
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/test/integration.TestBasicFlowsWithOriginatingIdentity()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/test/integration/controller_test.go:946 +0x29f6
[Build & Unit/Integration Tests]   testing.tRunner()
[Build & Unit/Integration Tests]       /usr/local/go/src/testing/testing.go:746 +0x16c
[Build & Unit/Integration Tests] 
[Build & Unit/Integration Tests] Previous read at 0x00c424c9ac68 by goroutine 1145:
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/apiserver/pkg/util/feature.(*featureGate).Enabled()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/apiserver/pkg/util/feature/feature_gate.go:182 +0x13a
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/pkg/controller.(*controller).reconcileServiceInstanceCredential()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/pkg/controller/controller_binding.go:275 +0x4ad9
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/pkg/controller.(*controller).reconcileServiceInstanceCredentialKey()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/pkg/controller/controller_binding.go:66 +0x2d8
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/pkg/controller.(*controller).(github.com/kubernetes-incubator/service-catalog/pkg/controller.reconcileServiceInstanceCredentialKey)-fm()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/pkg/controller/controller.go:178 +0x55
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/pkg/controller.worker.func1.1()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/pkg/controller/controller.go:209 +0xf7
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/pkg/controller.worker.func1()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/pkg/controller/controller.go:226 +0xb2
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:97 +0x6f
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:98 +0xcd
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/apimachinery/pkg/util/wait.Until()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:52 +0x5a
[Build & Unit/Integration Tests] 
[Build & Unit/Integration Tests] Goroutine 1034 (running) created at:
[Build & Unit/Integration Tests]   testing.(*T).Run()
[Build & Unit/Integration Tests]       /usr/local/go/src/testing/testing.go:789 +0x568
[Build & Unit/Integration Tests]   testing.runTests.func1()
[Build & Unit/Integration Tests]       /usr/local/go/src/testing/testing.go:1004 +0xa7
[Build & Unit/Integration Tests]   testing.tRunner()
[Build & Unit/Integration Tests]       /usr/local/go/src/testing/testing.go:746 +0x16c
[Build & Unit/Integration Tests]   testing.runTests()
[Build & Unit/Integration Tests]       /usr/local/go/src/testing/testing.go:1002 +0x521
[Build & Unit/Integration Tests]   testing.(*M).Run()
[Build & Unit/Integration Tests]       /usr/local/go/src/testing/testing.go:921 +0x206
[Build & Unit/Integration Tests]   main.main()
[Build & Unit/Integration Tests]       github.com/kubernetes-incubator/service-catalog/test/integration/_test/_testmain.go:70 +0x1d3
[Build & Unit/Integration Tests] 
[Build & Unit/Integration Tests] Goroutine 1145 (running) created at:
[Build & Unit/Integration Tests]   github.com/kubernetes-incubator/service-catalog/pkg/controller.(*controller).Run()
[Build & Unit/Integration Tests]       /go/src/github.com/kubernetes-incubator/service-catalog/pkg/controller/controller.go:178 +0x6d6
[Build & Unit/Integration Tests] ==================
[Build & Unit/Integration Tests] --- FAIL: TestBasicFlowsWithOriginatingIdentity (4.95s)
[Build & Unit/Integration Tests] 	framework.go:83: Starting server on port: 14945
[Build & Unit/Integration Tests] 	controller_test.go:1001: controller start
[Build & Unit/Integration Tests] 	controller_test.go:1009: informers start
[Build & Unit/Integration Tests] 	framework.go:79: Shutting down server on port: 14945
[Build & Unit/Integration Tests] 	testing.go:699: race detected during execution of test
[Build & Unit/Integration Tests] FAIL
```

### PR DESCRIPTION
**Pull request by staebler**
Fixes #1278.

* Modify controller.Run so that it waits for all worker goroutines to end before returning.
* Modify controller integration tests so that they wait for controller shutdown prior to finishing. This is most important for the test that tests for OriginatingIdentity because that test needs to wait for the controller workers to completely stop prior to disabling the OriginatingIdentity feature.

### CODE REVIEW

File: pkg/controller/controller.go [hunk 1]
```diff
@@ -188,6 +197,20 @@ func (c *controller) Run(workers int, stopCh <-chan struct{}) {
...
+
+	for i := 0; i < totalNumberOfWorkers; i++ {
+		<-done
+	}
+}
+
+// createWorker creates and runs a worker thread that just processes items in the
+// specified queue. The worker will run until stopCh is closed. When the worker
+// stops, a token will be added to the done channel.
+func createWorker(queue workqueue.RateLimitingInterface, resourceType string, maxRetries int, forgetAfterSuccess bool, reconciler func(key string) error, stopCh <-chan struct{}, done chan<- struct{}) {
```
Comment (MHBauer) [inline pkg/controller/controller.go]: Why an internal goroutine vs calling `go createWorker` ?

File: pkg/controller/controller.go [hunk 2]
```diff
@@ -170,13 +170,22 @@ func (c *controller) Run(workers int, stopCh <-chan struct{}) {
...
+		createWorker(c.serviceClassQueue, "ServiceClass", maxRetries, true, c.reconcileServiceClassKey, stopCh, done)
+		totalNumberOfWorkers++
+		createWorker(c.servicePlanQueue, "ServicePlan", maxRetries, true, c.reconcileServicePlanKey, stopCh, done)
+		totalNumberOfWorkers++
+		createWorker(c.instanceQueue, "ServiceInstance", maxRetries, true, c.reconcileServiceInstanceKey, stopCh, done)
+		totalNumberOfWorkers++
+		createWorker(c.bindingQueue, "ServiceInstanceCredential", maxRetries, true, c.reconcileServiceInstanceCredentialKey, stopCh, done)
+		totalNumberOfWorkers++
+		createWorker(c.pollingQueue, "Poller", maxRetries, false, c.requeueServiceInstanceForPoll, stopCh, done)
+		totalNumberOfWorkers++
```
Comment (MHBauer) [inline pkg/controller/controller.go]: Can this for loop fail partway through?

I'm not sure we need to have a variable and count up vs just calculating the total worker count directly.

Comment (staebler) [reply pkg/controller/controller.go]: My reasoning for counting was that the alternative required having a magic number for the count of queues. It seems to me to be pretty easy for that number to diverge from the real count. This way seemed less fragile from a maintainability point.

File: pkg/controller/controller.go (hunk 1 above)
Comment (staebler) [reply pkg/controller/controller.go]: No reason other than that it was more code at the caller side. Is one preferred over the other?

Comment (pmorie) [reply pkg/controller/controller.go]: I think the fix is correct, but should be implemented with [`WaitGroup`](https://golang.org/pkg/sync/#WaitGroup).

Ideally, there would be a variant of `wait.Until` that took a WaitGroup to call `Add` and `Done` on.  Then in `Run` you could just call `Wait`

**Comment by MHBauer** [comment]
https://travis-ci.org/kubernetes-incubator/service-catalog/builds/280724977 hey look it's #1275

File: pkg/controller/controller.go [hunk 3]
```diff
@@ -188,6 +191,19 @@ func (c *controller) Run(workers int, stopCh <-chan struct{}) {
...
 	c.bindingQueue.ShutDown()
 	c.pollingQueue.ShutDown()
+
+	waitGroup.Wait()
+}
+
+// createWorker creates and runs a worker thread that just processes items in the
+// specified queue. The worker will run until stopCh is closed. The worker will be
+// added to the wait group when started and marked done when finished.
+func createWorker(queue workqueue.RateLimitingInterface, resourceType string, maxRetries int, forgetAfterSuccess bool, reconciler func(key string) error, stopCh <-chan struct{}, waitGroup *sync.WaitGroup) {
```
Comment (MHBauer) [inline pkg/controller/controller.go]: My thought on the goroutine above vs internally is that this is `createAndStartWorker`. Not obvious in the use context. The `go` at the calling site says "we're spawning threads in a loop" in an idiomatic way that is immediately apparent. 

The wait group sort of alerts you to it. 

I also want to say we're getting into having way to many arguments to a function, but it's probably okay for now.

**Pull request merged**